Fix text save bug and add file load feature.

10carfernandez opened the pull request:
Separate code for saving texts and screenshots into its own file,
FileManager, which also contains code for loading files.

I was considering changing the "PIC" folder to "SAVES," since the stuff under "PIC" also includes non-picture files. 

It works the way you'd expect it to work - if you hit "Load File," a window pops up and you can choose which text file that was previously saved to load. 

I also included a "Change Save Folder" button that lets you choose or create a new folder in which you can save the text and pic files during that current session. 

What will happen is, whatever folder you choose as your new save directory will have each of the two text and screenshot folders created automatically. 

Speaking of buttons, the buttons were changed. After adding 2 new buttons, they looked all out of order, so I rearranged them so that they'd look better. I will attach a screenshot. This was a real pain in the neck, so I might consider making a separate file that deals just with the UI in the future.

10carfernandez commented:
![capture](https://cloud.githubusercontent.com/assets/21129921/22050561/90116a50-dd01-11e6-90c4-2e787fe6d72f.JPG)

PRabahy commented on src/evolv/io/EvolvioColor.java:
```diff
@@ -58,60 +68,118 @@ public void settings() {
+		FileManager fileManager= new FileManager(this, INITIAL_FILE_NAME);
+
+		// Open file and load object
+		evoBoard = fileManager.fileLoad(fullPath);
```
Looks like fileLoad can return null, so this could end up throwing a null reference exception.

PRabahy commented on src/evolv/io/FileManager.java:
```diff
@@ -0,0 +1,101 @@
+	}
+	
+	void prepareForFileSave(int type) {
+		fileSaveTimes[type] = -999999;
```
I'm not a fan of magic numbers like this. I would rather see Integer.MAX_VALUE or Integer.MIN_VALUE depending on what is happening.

PRabahy commented on src/evolv/io/FileManager.java:
```diff
@@ -0,0 +1,101 @@
+	// Save entire board or image of frame
+	void fileSave(Board evoBoard) {		
+		for (int i = 0; i < 4; i++) {
+			if (fileSaveTimes[i] < -99999) {
```
If this is referring to the "same" -99999 as line 78, please turn it into a constant.

10carfernandez replied:
It's not the same as the other one, but I think it was mean to be. So I'm going ahead and setting a constant MIN_VALUE to Integer.MIN_VALUE for those two values.

10carfernandez replied:
OK it looks like each of those -99999 values are distinct and are being used to kill several birds with one stone. When they become the value year, the value is used to calculate the next filesave. 

I'm creating a boolean array that will accomplish the same thing and make the code more readable. I don't like the use of these magic numbers either, and it doesn't seem to make such a huge difference in performance (especially if we were to turn those numbers into constants anyways).

PRabahy commented on src/evolv/io/SoftBody.java:
```diff
@@ -139,6 +143,7 @@ public void applyMotions(double timeStep) {
 
 	public void drawSoftBody(float scaleUp) {
 		double radius = getRadius();
+		this.evolvioColor = board.evolvioColor;
```
Why this new line? this.evolvioColor was already set in the constructor.

10carfernandez replied:
Whoops, I forgot to take that off.

PRabahy commented on src/evolv/io/SoftBody.java:
```diff
@@ -49,6 +52,7 @@ public SoftBody(EvolvioColor evolvioColor, double tpx, double tpy, double tvx, d
 		saturation = tsaturation;
 		brightness = tbrightness;
 		board = tb;
+		this.evolvioColor = tb.evolvioColor;
```
Why did this line need to move?

10carfernandez replied:
Oh, that line should have been this.evolvioColor = _board_.evolvioColor, but it's probably better if I leave it back where it was an append the _tb_ to it so that it's this.evolvioColor = _tb_.evolvioColor.

PRabahy commented on src/evolv/io/Board.java:
```diff
@@ -4,11 +4,15 @@
+	/**
+	 * 
+	 */
+	public EvolvioColor evolvioColor;
```
If we are going to make this public, lets just choose one class to store it. That would eliminate the need to pass it in a bunch of constructors.

10carfernandez replied:
I'll go ahead and change that then.

10carfernandez replied:
Wasn't there a discussion about gradually moving away from the processing library? I ask because the NameGenerator file uses EvolvioBoard but only to generate random numbers. I was thinking about taking out the constructor and using Math.random() instead.

PRabahy replied:
If that's all NameGenerator uses the reference to evolvioColor for, I agree we should change those to java.util.Random, but its probably outside the scope of this PR (unless it makes this implementation easier).

10carfernandez replied:
I think it might help in the future if/when we decide to start saving brains and creatures separately, since it now gets evolvioColor from Creature, which gets it from Board. Even though EvolvioColor is serializable, the PApplet isn't.

Either way, I think this will probably come up in a future PR when we start saving "strains" of creatures and brains. So it's probably better to leave it for that, instead of doing a ton of changes in a single PR.

PRabahy commented:
I was just trying to test this and found that the change to the .classpath file broke the build. It sounds like the .classpath shouldn't even be included in the repo. I'll fix the gitignore file to make this easier/better.

PRabahy commented:
If you rebase this on top of #87 it will fix the build.

10carfernandez commented:
Sorry, I'm a huge noob to git. I know what I have to do, but don't know how to and couldn't find help. So I just manually updated the .gitignore file. Was there a simple way of rebasing this with what you did?

PRabahy commented:
@10carfernandez Rebasing is a little weird, but here is what I would do.

1. Get my PR
1a. git checkout -b PRabahy-feature/ImproveGitIgnore master (creates a local branch on your computer for my changes)
1b. git pull https://github.com/PRabahy/evolv.io.git feature/ImproveGitIgnore (actually gets my changes from github)
2. Switch back to your branch.
2a. git checkout master (I believe you have been developing on your master branch)
3. Do the actual rebase.
3a. git rebase feature/ImproveGitIgnore (You are rebasing onto the branch you made in step 1)
4. Resolve any merge conflicts. I believe the only conflict will be the .classpath file. Just make sure it gets deleted as the resolution.
5. Push the changes to github
5a. git push -f (You will need to force push because a rebase rewrites history).

If all that is too scary, I can do all the fixups on my PC, push to master, then we can close this PR. It won't take me much time.

I don't want to get too hung up on this detail. I would rather see more people reviewing the concept of the PR and testing that it works without bugs.

PRabahy commented:
Cool, I was able to test it a little bit and fundamentally it appears to work. I noticed a couple of things.

1. The Load File button uses the native file browser, but Change Save Folder uses the Java one.
2. The files save as .txt, which tries to open in notepad. I would tend to save them as something more like .evolve.

10carfernandez commented:
I should also mention that I changed the file "PIC" to "SAVES." So that might have be updated in the .gitignore PR you made. Alternatively, I can change it back to "PIC" if you think I should leave it like that. 

In that same commit, I updated the code so that all the save folders that would be saved in the same directory as "PIC" are now saved under "SAVES" (or "PIC") as "default" or as whatever you want to name it. That way, you don't have to worry about new folders not being ignored when using git. 

1. I noticed this as well. This seems to be how the PApplet library wants to do it for selecting a folder, as opposed to just a file.  I figured this would do for now until we implemented a start screen, as mentioned in #82 . Once or if we implement that, saving files would just consist of giving them a name, and loading them would consist of just choosing from a list of names of previously saved simulation sessions.  

2. OK I changed the extension to ".evolve" and decided to go ahead and update the other info (like autoTexts to just autoSave) to keep everything more consistent.

NateUniAccount commented:
This seems to be a very large commit, is there any chance of breaking it down into smaller commits so it's easier to see what's going on?

10carfernandez commented:
Sorry for the late reply. 

1. I'm not sure what unrelated things you're talking about. Was there anything specific that stood out to you that seemed unrelated? 
2. I can try doing that. Sorry if my changes are slow/sluggish/bad.

NateUniAccount commented:
Sure of course, here's what stood out to me as unrelated to file saving/loading:

In Board there's an extra columns field
In Board drawUI() the button texts are in a different order and the case statements are changed
In Board toBigString() there's an extra unused line
In Brain, Creature, SoftBody and Tile they reference a Board instead of an EvolvioColor and often just delegate to the EvolvioColor
In EvolvioColor mousePressed() the case statements are in a different order


Oh no not at all! :) They look useful, it's just there's too much for a single commit that's all. It means it takes ages to check it all and make sure there's no breaking changes, and people will be less inclined to say it's ok in case they miss something.
If it's broken up into smaller commits, it's much easier to see whether it's ok or if something's wrong, people will be much happier to say it's ok and it'll get added a lot quicker!

10carfernandez commented:
Are you saying that the UI changes I made should be done in a different PR? 

I can't think of a way of testing the file save/load methods without creating the extra buttons to test those methods. 

I _did_ try making parts of the UI more scalable when I was adding the new buttons because I noticed that there are a lot of hard-coded values for the UI, which meant that if you tried adding an extra button, extra columns of buttons, or moving the buttons, the correct methods would no longer get called (e.g., pressing on "save text" would actually call the "save image" method). 
Those changes were the ones done to:

> In Board there's an extra columns field
> In Board drawUI() ... the **case statements are changed**
> In EvolvioColor mousePressed() the case statements are in a different order

So I can probably apply those changes in a separate PR and have the new buttons I included be part of this PR. The other problem is that adding the extra columns to fit the new buttons in the UI cause the buttons to look way out of order, so I had to re-order the buttons. 

The other changes you mentioned might be very difficult to do in a separate PR because of all the inter-dependencies involved in making the objects and simulation session serializable and because of the fact that  the Processing objects aren't serializable (that's why I had to have the other objects reference Board, instead of EvolvioColor, since the latter extends a Processing object). 
And the changes I'm referring to specifically are:

> In Brain, Creature, SoftBody and Tile they reference a Board instead of an EvolvioColor and often just delegate to the EvolvioColor
> In Board drawUI() the **button texts are in a different order**...

I'm not sure what happened with the toBigString() method.  I'll change it back to how it was before. 

I can definitely break them up into smaller commits, though. I'm still trying to get used to this type of source control, so it might take me a while.

NateUniAccount commented:
What I'm asking is for smaller commits, there's 850 lines that have been changed, from a reviewer's point of view it's just too much to review and make sure there's no unwanted side effects. How it's broken up doesn't really matter, functionality can be added before it's used that's no problem. It does look good :) it just needs a few smaller commits that's all
